### ISSUE TITLE
[operator] watch for changes to remote cluster secrets

### ISSUE THREAD

**Issue by jmazzitelli** [kiali/kiali#6941]
Right now in order to get the Kiali Server to auto-discover new remote cluster secrets or changes to existing ones, the operator has to be poked (manually update the Kiali CR so the operator runs a reconcile).

This is not optimal.

We should consider adding a watch to the [Ansible watches.yaml](https://sdk.operatorframework.io/docs/building-operators/ansible/reference/watches/) so the operator will automatically reconcile when it detects new secrest coming online or old secrets getting deleted. The operator can watch secrets that match the default label [kiali.io/multiCluster=true](https://github.com/kiali/kiali-operator/blob/v1.77.0/roles/default/kiali-deploy/defaults/main.yml#L252). (Side note: in order for this automatic reconciliation to work, the user must not change that default value in their Kiali CR. If for some reason they do, they can always fallback to the manual "poke" to get the reconcile to work. But it will be rare for people to want to change that label).

**Comment by jmazzitelli** [comment]
> The operator can watch secrets that match the default label [kiali.io/multiCluster=true](https://github.com/kiali/kiali-operator/blob/v1.77.0/roles/default/kiali-deploy/defaults/main.yml#L252). (Side note: in order for this automatic reconciliation to work, the user must not change that default value in their Kiali CR. If for some reason they do, they can always fallback to the manual "poke" to get the reconcile to work. But it will be rare for people to want to change that label

This will fix that: https://github.com/operator-framework/ansible-operator-plugins/pull/45
In this case if they want to label their remote cluster secrets differently, they can tell the operator about it by setting some env variables on the operator. This requires us to pick up the new operator base image with that code - no idea when that will be available.

**Comment by jmazzitelli** [comment]
The [enhancement](https://github.com/operator-framework/ansible-operator-plugins/pull/45) to the ansible operator base image has been merged. We would have to wait for it to make it to release and for Kiali operator to be able to upgrade its base image before we can utilize it.

**Comment by jmazzitelli** [comment]
(FYI: PR https://github.com/kiali/kiali/issues/7877 is somewhat related to this one.)

**Comment by jmazzitelli** [comment]
> Is a change to https://github.com/kiali/kiali-operator/blob/master/watches-k8s.yaml required to make this work?

For automatic detection, yes, I think the only changed would need to be made to the two watches.yaml files ([watches-k8s.yaml](https://github.com/kiali/kiali-operator/blob/master/watches-k8s.yaml) and [watches-os.yaml](https://github.com/kiali/kiali-operator/blob/master/watches-os.yaml)).

Since [the operator already has cluster "watch" permission on secrets](https://github.com/kiali/helm-charts/blob/master/kiali-operator/templates/clusterrole.yaml#L32-L38), I don't think it needs to be given any more permissions. But I haven't looked into it to know if there is anything else that is needed.

The new watch will need to be on secrets with that specific label using a selector:

```yaml
...
  selector:
    matchLabels:
      kiali.io/multiCluster: "true"
```

**NOTE: the below is a long set of details that, in the end, I don't think we want to do. I'm just documenting it here for future reference. Read the last few sentences in this comment down below :)**

---

But because the Kiali CR allows for [a custom label to be defined](https://github.com/kiali/kiali-operator/blob/v2.6/roles/default/kiali-deploy/defaults/main.yml#L46) we might want to use an env var here, as this issue's description mentioned. That [feature](https://github.com/operator-framework/ansible-operator-plugins/pull/45) we've been needing went into [upstream ansible operator base image 1.34.0](https://github.com/operator-framework/ansible-operator-plugins/releases/tag/v1.34.0), so it is available for us to use (the base image we are using for kiali operator is newer so it has that feature).

 Something like this:

```yaml
...
  selector:
    matchLabels:
      ${REMOTE_CLUSTER_SECRET_LABEL_NAME}: "${REMOTE_CLUSTER_SECRET_LABEL_VALUE}"
```

Then just define those env vars with their default values in the operator templates, after the WATCHES_FILES env var definition in [kiali-ossm manifest](https://github.com/kiali/kiali-operator/blob/v2.6.0/manifests/kiali-ossm/manifests/kiali.clusterserviceversion.yaml#L351), [kiali-upstream manifest](https://github.com/kiali/kiali-operator/blob/master/manifests/kiali-upstream/2.6.0/manifests/kiali.v2.6.0.clusterserviceversion.yaml#L273), and [helm chart template](https://github.com/kiali/helm-charts/blob/v2.6.0/kiali-operator/templates/deployment.yaml#L131-L140).

The annoying thing here is the user will need to ensure ALL their Kiali installations use the same label, and have all the Kiali CRs use that custom label (because the operator can only have one env var). But, honestly, I don't see why anyone would want to change the label from the default anyway; however, my experience in the industry tells me that SOMEONE SOMEWHERE is going to want to do it, so that's why the customization is there. We could always tell that user "if you want to use a custom label on your cluster secrets, then you lose auto-reconciliation ability in the operator - in that case, just poke the CR after you update the secrets to get the operator to reconcile it". If we elect to do that, all that additional work with the env var can just be skipped. We could just update the watches file.

---

As I typed all that, I think I'm of the opinion we just change the watches file. If for some reason customizing that label is a common want for users, we can circle back and add the env var stuff. I've documented all the work that needs to be done above, so it would be simple for someone to come in and retrofit the operator with those env vars.

In short, yes, just change the two watches files :)

**Comment by jmazzitelli** [comment]
There are some issues with the proposed solution. I tried what I thought would work in this PR:

* https://github.com/kiali/kiali-operator/pull/891

However, 2 issues arose:

1. What I found is that the operator controller needs not only "watch" permission but also "get" (which makes sense now that I think about it - we are going to need the secret's metadata to obtain things like its namespace and managedFields (for its modified timestamp)). The operator today does not have global "get" permission for secrets in the cluster, and this results:
```
{"level":"error","ts":"2025-03-18T20:24:45Z","msg":"Reconciler error","controller":"secret-controller","object":
{"name":"foo","namespace":"kiali-operator"},"namespace":"kiali-operator","name":"foo","reconcileID":
"3e1faa5f-af14-4722-bb87-2b9e00ad996b","error":"secrets \"foo\" is forbidden: User
\"system:serviceaccount:kiali-operator:kiali-operator\" cannot get resource \"secrets\" in API group \"\"
in the namespace \"kiali-operator\"","stacktrace":...
```
I am wary about adding global "get" secret perms, due to the security issues involved in allowing the operator role to "get" secrets cluster wide.

2. In addition, we don't want to reconcile all Kiali CRs in all namespaces when a secret is touched. We only want to reconcile those whose Kiali servers are in the same namespace as the secret (because only those servers can see the secret anyway - you can't bind a secret to a volume in a pod that lives in a different namespace). I originally implemented the playbook so it will only reconcile those Kiali CRs that are in the same namespace as the secret. But that is not good enough because it is possible to have the Kiali CR in one namespace where the server is deployed in another namespace (that's what the Kiali CR `spec.deployment.namespace` allows for). So if the server and secret are in `istio-system`, but the Kiali CR is in `kiali-operator`, that playbook will be a no-op. The secret will be in `istio-system` and the operator will therefore look for the Kiali CR in `istio-system`, but it isn't there. At minimum we would have to document that this watching of secrets and auto-detecting them via the operator will only work if the user puts the Kiali CR in the same namespace as where the Kiali Server is to be deployed. That is the typical use-case, so this isn't a major concern, but it is a concern because while having the CR in a different namespace from the server is unusual, it is a valid use case. So this limitation would need to be documented.

Of course, the important thing is the "get" secret permission. I'm not sure about that.

**Comment by jmazzitelli** [comment]
To mitigate the requirement of the operator given "get" secret perms cluster wide, we could implement [this](https://github.com/kiali/kiali/issues/7877) which will allow multiple remote secrets to be specified in a single secret. We could then require that secret be named specifically something like "kiali-remote-cluster-secret" and then the operator would be limited to reading only those secrets cluster wide with that name:

```
- apiGroups: [""]
  resourceNames:
  - kiali-remote-cluster-secret
  resources:
  - secrets
  verbs:
  - get
  - list
  - watch
```

As long as the user names their secret with that name (and labels it with `kiali.io/multiCluster: true`) then the operator would be able to successfully watch for them.

Also, that [issue](https://github.com/kiali/kiali/issues/7877) makes the good point that if we have a multi-cluster single secret, we could mount that to the kiali pod optionally - so the kiali pod will always attempt to mount it upon creation. In this case, the operator would merely have to rollout a new kiali server deployment (i.e. kill the pod and let it restart) and it should pick up the new/updated secret). So the playbook wouldn't need to perform a full reconciliation - it would only need to kill the kiali pod and it will restart with the updated secret.

**Comment by jmazzitelli** [comment]
Some docs will be needed to point this feature out - this blurb might be all that we need: https://github.com/kiali/kiali.io/pull/868#issuecomment-2738097093

**Comment by jmazzitelli** [comment]
PRs:

* operator: https://github.com/kiali/kiali-operator/pull/891
* helm: https://github.com/kiali/helm-charts/pull/316
* docs: https://github.com/kiali/kiali.io/pull/869

**Comment by jshaughn** [comment]
> we could implement https://github.com/kiali/kiali/issues/7877

Not my wheelhouse, but this approach sounds good to me.

**Comment by aghassabian** [comment]
@jmazzitelli Shouldn't this line to be updated to match the `kiali.io/kiali-multi-cluster-secret: "true"` label?

**Comment by aghassabian** [comment]
@jmazzitelli What's your opinion of updating [kiali-prepare-remote-cluster.sh](https://github.com/kiali/kiali/blob/master/hack/istio/multicluster/kiali-prepare-remote-cluster.sh) script with a new flag to support creating `kiali-multi-cluster-secret` with `kiali.io/kiali-multi-cluster-secret` label?

**Comment by jmazzitelli** [comment]
> What's your opinion of updating [kiali-prepare-remote-cluster.sh](https://github.com/kiali/kiali/blob/master/hack/istio/multicluster/kiali-prepare-remote-cluster.sh) script with a new flag to support creating `kiali-multi-cluster-secret` with `kiali.io/kiali-multi-cluster-secret` label?

Feel free to write up a github enhancement request issue and it can be discussed there: https://github.com/kiali/kiali/issues/new?template=feature_request.md

It won't be as easy at first glace. That script's implementation and options all assumes a single remote cluster. And it uses the server helm chart under the covers which produces the remote resources on the remote cluster as an optional feature. It's possible but it isn't trivial. You would have to have a way to specify and operate on multiple remote clusters while maintaining backwards compatibility with the current single cluster options.

See:
* https://github.com/kiali/kiali/blob/v2.12.0/hack/istio/multicluster/kiali-prepare-remote-cluster.sh#L501-L520
